GCS: Cleanup UAVObjectWidgetUtils

tracernz opened the pull request:
I've had this sitting around for weeks so time to get it upstream before it rots.

- Use .ui files for UAVO relations wherever appropriate.
  - Cleanup a few things relating to these.
- Remove most of the addUAVObjectToWidgetRelation overloads
- Move some of the public API to private to prevent it making it's way back in
- Cleanup includes a little

tracernz committed:
GCS: Cleanup UAVObjectWidgetUtils

mlyle commented on ground/gcs/src/plugins/sysalarmsmessaging/sysalarmsmessagingplugin.cpp:
```diff
@@ -93,7 +93,7 @@ void SysAlarmsMessagingPlugin::updateAlarms(UAVObject *systemAlarm)
 {
     UAVObjectField *field = systemAlarm->getField(QString("Alarm"));
 
-    for (uint i = 0; i < field->getNumElements(); ++i) {
```
Is this use of 'auto' a win for loop iterators?   In this kind of use it's synonymous with 'int' (e.g. take the type of 0, ignore the other use types).

I mean I guess it's OK because 32 bits is almost always wide enough-- though there is also the whole signed vs. unsigned compare issue.

tracernz replied:
The signed/unsigned issue is why I liked auto... I have moved everything to int though so happy to make the iterator vars ints if desired.

mlyle replied:
Yah, auto just looks at the initializer though, right?  and `0` is an int (as opposed to `0U` etc) so this is functionally equivalent to int, isn't it?

If they're all int now it's immaterial, but I don't think it's a preferable style.

tracernz replied:
You're right, I'll change it (and one other place too I think). :+1:

mlyle commented on ground/gcs/src/plugins/uavobjectwidgetutils/configtaskwidget.cpp:
```diff
@@ -1189,14 +1162,13 @@ void ConfigTaskWidget::disconnectWidgetUpdatesToSlot(QWidget *widget, const char
                                           double scale, bool usesUnits)
 {
-    if (!widget || !field)
+    if (!widget || !field || widgetReadOnly(widget))
```
nice

mlyle commented on ground/gcs/src/plugins/uavobjectwidgetutils/configtaskwidget.h:
```diff
@@ -28,27 +28,30 @@
 #include <QUrl>
 #include <QEvent>
 
+class UAVObject;
```
Thank you.  Have hated just how much we suck in.

mlyle reviewed (approved):
All good stuff.

mlyle commented:
My one concern is this-- if something is newly disconnected from the UAVO, my skimming of the code is unlikely to have spotted it.

Have you been able to use your clang tools to mechanically verify names/connections?

No matter the answer to this, I fear this changeset requires a fair bit of manual test, whether pre-merge or post-merge-- to rule out regression.

tracernz commented:
> Have you been able to use your clang tools to mechanically verify names/connections?

Not yet, but this is a great reason for me to tidy up and bring that work upstream too. 👍

tracernz committed:
Gcs:SysAlarmsMessaging: Fixes from review feedback

tracernz commented:
For reference, this is the script I used on the OSD widget UI code, which did most of the work for me, and took the mistakes out ;):
```python
#!/usr/bin/env python3 
 
import sys 
 
from lxml import etree 
 
def main(): 
    uifile = sys.argv[1] 
    uavofile = sys.argv[2] 
 
    with open(uifile) as fp: 
        uitree = etree.parse(uifile) 
    with open(uavofile) as fp: 
        uavotree = etree.parse(uavofile) 
 
    fields = [] 
    for f in uavotree.xpath('//field'): 
        fields.append(f.get('name')) 
 
    for sb in uitree.xpath('//widget[@class="QSpinBox"]') + uitree.xpath('//widget[@class="QComboBox"]'): 
        name = sb.get('name') 
        ucname = name[0].upper() + name[1:] 
        field = ucname 
        if name.endswith('X'): 
            field = ucname[:-1] + 'PosX' 
        elif name.endswith('Y'): 
            field = ucname[:-1] + 'PosY' 
        if field not in fields: 
            field = ucname[:-1] + 'Pos' 
        if field not in fields: 
            print(name) 
            continue 
 
        relation = etree.Element('property', {'name': 'objrelation', 'stdset': '0'}) 
        strlist = etree.Element('stringlist', {'notr': 'true'}) 
        s = etree.Element('string') 
        s.text = 'fieldname:' + field 
        strlist.append(s) 
        relation.append(strlist) 
 
        for oldrel in sb.xpath('property[@name="objrelation"]'): 
            oldrel.getparent().remove(oldrel) 
 
        sb.append(relation) 
 
    for cb in uitree.xpath('//widget[@class="QCheckBox"]'): 
        name = cb.get('name') 
        ucname = name[0].upper() + name[1:] 
        field = ucname 
        if name.endswith('Enabled'): 
            field = ucname[:-len('Enabled')] 
        if field not in fields: 
            print(name) 
            continue 
 
        relation = etree.Element('property', {'name': 'objrelation', 'stdset': '0'}) 
        strlist = etree.Element('stringlist', {'notr': 'true'}) 
        s = etree.Element('string') 
        s.text = 'fieldname:' + field 
        strlist.append(s) 
        s = etree.Element('string') 
        s.text = 'checkedoption:Enabled' 
        strlist.append(s) 
        s = etree.Element('string') 
        s.text = 'uncheckedoption:Disabled' 
        strlist.append(s) 
        relation.append(strlist) 
 
        for oldrel in cb.xpath('property[@name="objrelation"]'): 
            oldrel.getparent().remove(oldrel) 
 
        cb.append(relation) 
 
    newfile = '.'.join(uifile.split('.')[:-1]) + '.fixed.xml' 
    with open(newfile, 'wb') as fp: 
        fp.write(etree.tostring(uitree, pretty_print=True, xml_declaration=True, encoding='UTF-8')) 
 
if __name__ == '__main__': 
    sys.exit(main()) 
```

tracernz commented:
Ran my script that checks the uavo relations in .ui files (with a couple of checks disabled because they're currently noisy), found no issues from this PR, but some other stuff on next is broken:
```
In file: /home/mike/dev/dronin2/ground/gcs/src/plugins/config/autotuneproperties.ui
    Error: measuredRollNoise [QLabel]: Invalid field SystemIdent.Noise
    Error: measuredYawNoise [QLabel]: Invalid field SystemIdent.Noise
    Error: measuredYawBias [QLabel]: Invalid field SystemIdent.Bias
    Error: measuredYawGain [QLabel]: Invalid field SystemIdent.Beta
    Error: measuredRollGain [QLabel]: Invalid field SystemIdent.Beta
    Error: measuredPitchNoise [QLabel]: Invalid field SystemIdent.Noise
    Error: measuredRollBias [QLabel]: Invalid field SystemIdent.Bias
    Error: measuredPitchBias [QLabel]: Invalid field SystemIdent.Bias
    Error: measuredPitchGain [QLabel]: Invalid field SystemIdent.Beta
In file: /home/mike/dev/dronin2/ground/gcs/src/plugins/boards_brotronics/luxconfiguration.ui
    Error: help [QPushButton]: HTTP Error 404: Not Found
```

tracernz committed:
GCS:Config:OSD: Add missing notr to button UAVO relations

mlyle commented:
Very nice :)
